File: katas/5 kyu/find-the-unique-string/solution.py
import codewars_test as test
# -------------------------------- my solution ------------------------------- #
def find_uniq(arr):
  for x in range(0,len(arr)):
    c = set(sorted(arr[x-1].lower()))
    a = set(sorted(arr[x].lower()))
    b = set(sorted(arr[x+1].lower()))
    if a != b:
      if a!=c:
        return arr[x]
      else:
        return arr[x+1]
# ------------------------------ clever solution ----------------------------- #
def find_uniq(arr):
    if set(arr[0].lower()) == set(arr[1].lower()):
        majority_set = set(arr[0].lower())
    elif set(arr[0].lower()) == set(arr[2].lower()):
        majority_set = set(arr[0].lower())
    else:
        majority_set = set(arr[1].lower())
    
    for string in arr:
        if set(string.lower()) != majority_set:
            return string
# ----------------------------------- test ----------------------------------- #
test.describe('should handle sample cases')
test.assert_equals(find_uniq([ 'Aa', 'aaa', 'aaaaa', 'BbBb', 'Aaaa', 'AaAaAa', 'a' ]), 'BbBb')
test.assert_equals(find_uniq([ 'abc', 'acb', 'bac', 'foo', 'bca', 'cab', 'cba' ]), 'foo')
test.assert_equals(find_uniq([ '    ', 'a', '  ' ]), 'a')

File: katas/5 kyu/find-the-unique-string/solution.php
<!-- find-the-unique-string -->
<?php
use PHPUnit\Framework\TestCase;
/* ------------------------------- MY SOLUTION ------------------------------ */
function find_uniq($arr) {

  if (in_array("" , $arr) || in_array(" " , $arr)){
    $arr = array_filter( $arr , fn($x) => $x != "" && $x != " "); return end($arr);
  }
  
  $cn = function(String $str) { $r = array_unique(str_split(strtolower($str))) ; sort($r) ; return implode("" , $r);};

  $s = $w = array_map($cn , $arr);
  sort($s);
  $ndx = $s[0] == $s[1] ? array_keys($w , end($s)) : array_keys($w , $s[0]);
  return $arr[$ndx[0]];
}
/* ----------------------------- CLEVER SOLUTION ---------------------------- */
/* ---------------------------------- TEST ---------------------------------- */
class FindUniqTest extends TestCase
{
  // test function names should start with "test"
  public function testSampleCases() {
    $this->assertSame('BbBb', find_uniq([ 'Aa', 'aaa', 'aaaaa', 'BbBb', 'Aaaa', 'AaAaAa', 'a' ]));
    $this->assertSame('foo', find_uniq([ 'abc', 'acb', 'bac', 'foo', 'bca', 'cab', 'cba' ]));
    $this->assertSame('victor', find_uniq([ 'silvia', 'vasili', 'victor' ]));
    $this->assertSame('Harry Potter', find_uniq([ 'Tom Marvolo Riddle', 'I am Lord Voldemort', 'Harry Potter' ]));
    $this->assertSame('a', find_uniq([ '     ', 'a', ' ' ]));
  }
}